Zinc/Private/ZincRepo.m

```diff
@@ -1044,6 +1047,206 @@ - (ZincTask*) queueGarbageCollectTask
+    NSParameterAssert(bundleID);
+    NSParameterAssert(zincVersionString);
+
+    return bundleID;
+}
+
+- (NSString *)manifestNameForBundleName:(NSString *)bundleName {
+    return [bundleName stringByAppendingString:@".json"];
+}
+
+- (NSURL *)manifestsFolderURL {
+    return [self absoluteURLForPathRelativeToZincFolder:@"manifests" isDirectory:YES];
+}
+
+- (NSURL *)manifestURLForManifestName:(NSString *)manifestName {
+    return [[self manifestsFolderURL] URLByAppendingPathComponent:manifestName
+                                                      isDirectory:NO];
+}
+
+- (BOOL)deleteFileAtURL:(NSURL *)url {
+    NSError *error;
+    [NSFileManager.defaultManager removeItemAtURL:url error:&error];
+    if (error) {
+        NSLog(@"Error while deleting file at url: %@ \nfor file path: %@",
+              error.localizedDescription,
+              url.path);
+    }
+
+    return !error;
+}
+
+- (void)purgeBundlesWithIDs:(NSSet<NSString *> *)bundleIDs
+                 bundleURLs:(NSSet<NSURL *> *)bundleURLs {
+    @synchronized(self.index) {
+        for (NSString *bundleID in bundleIDs) {
+            [self postNotification:ZincRepoBundleWillDeleteNotification
+                          bundleID:bundleID];
+            [self.index purgeBundleWithID:bundleID];
+        }
+    }
+
+    ZincTaskRef* taskRef = [[ZincTaskRef alloc] init];
+    taskRef.completionBlock = ^{
+        for (NSURL *bundleURL in bundleURLs) {
+            NSString *bundleName = [self bundleNameForBundleURL:bundleURL];
+            NSString *manifestName = [self manifestNameForBundleName:bundleName];
+            NSURL *manifestURL = [self manifestURLForManifestName:manifestName];
+
+            if ([self deleteFileAtURL:manifestURL]) {
+                [self deleteFileAtURL:bundleURL];
+            } else {
+                NSLog(@"Failed to delete a manifest, will not delete the related bundle: %@",
+                      bundleURL);
+            }
+        }
+
+        [NSNotificationCenter.defaultCenter postNotificationName:kZincAllContentBundlesWereDeleted
+                                                          object:self
+                                                        userInfo:nil];
+    };
+    [taskRef addDependency:[self queueIndexSaveTask]];
+    [self.taskManager addOperation:taskRef];
+}
+
+- (void) deleteAllContentBundlesIfNeeded {
+    float totalSizeOfContentBundles = [self totalSizeOfContentBundles];
+    NSString *message = [NSString stringWithFormat:@"Total size of content bundles: %f", totalSizeOfContentBundles];
+
+    if (totalSizeOfContentBundles < kContentBundleFlushLimitInMegabytes) {
+        NSLog(@"%@, will not clear", message);
+        return;
+    }
+    NSLog(@"%@, will clear", message);
+
+    NSMutableSet<NSURL *> *contentBundleURLs = [NSMutableSet new];
+    NSMutableSet<NSString *> *contentBundleIDs = [NSMutableSet new];
+    [self forBundlesWithPrefix:kContentBundlePrefix performBlock:^(NSURL *bundleURL) {
```

Comment:
right, I'm definitely deleting from `repo.json` first and making sure it's successful before deleting the actual bundles and manifests.

I'm just suggesting that it doesn't have to match what actually in `repo.json` (what was actually deleted). It can just match the format `@"com.wonder.content"`, and that would be sufficient, if not even better.